model/host/host.go

```diff
@@ -1660,17 +1666,44 @@ func StaleRunningTaskIDs(staleness time.Duration) ([]task.Task, error) {
 // ModifySpawnHost updates a spawnhost with the changes described
 // in a HostModifyOptions struct.
 func (h *Host) ModifySpawnHost(opts HostModifyOptions) error {
-	// Delete instance tags
-	for _, key := range opts.DeleteInstanceTags {
-		delete(h.InstanceTags, key)
+	h.DeleteTags(opts.DeleteInstanceTags)
+	h.AddTags(opts.AddInstanceTags)
+	if err := h.SetTags(); err != nil {
+		return errors.Wrap(err, "error modifying spawn host")
+	}
+	return nil
+}
+
+// AddTags adds the specified tags to the host document, or modifies
+// an existing tag if it can be modified.
+func (h *Host) AddTags(tags []Tag) {
+	for _, new := range tags {
+		for i, old := range h.InstanceTags {
+			if old.Key == new.Key && old.CanBeModified {
+				h.InstanceTags[i] = new
+				break
+			}
+		}
+		h.InstanceTags = append(h.InstanceTags, new)
 	}
+}
 
-	// Add instance tags
-	for key, value := range opts.AddInstanceTags {
-		h.InstanceTags[key] = value
+// DeleteTags removes tags specified by their keys, only if those
+// keys are allowed to be deleted.
+func (h *Host) DeleteTags(keys []string) {
+	for _, key := range keys {
+		for i, tag := range h.InstanceTags {
+			if tag.Key == key && tag.CanBeModified {
+				h.InstanceTags[len(h.InstanceTags)-1], h.InstanceTags[i] = h.InstanceTags[i], h.InstanceTags[len(h.InstanceTags)-1]
+				h.InstanceTags = h.InstanceTags[:len(h.InstanceTags)-1]
```

Comment:
This is a bit complicated to read, and we don't usually assign to tuples. Can you do something like `	myList = append(myList[:i], myList[i+1:]...)` (this works also when i is the first/last index).